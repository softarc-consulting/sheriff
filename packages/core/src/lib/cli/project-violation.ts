import { Violation } from './verify';

export type ProjectViolation = {
  totalDependencyRuleViolations: number;
  totalEncapsulationViolations: number;
  totalViolatedFiles: number;
  hasError: boolean;
  violations: Violation[];
};
