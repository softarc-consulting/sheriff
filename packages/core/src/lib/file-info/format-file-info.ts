import UnassignedFileInfo from './unassigned-file-info';
import traverseUnassignedFileInfo from './traverse-unassigned-file-info';
import { EOL } from 'os';

export const formatFileInfo = (fileInfo: UnassignedFileInfo, indent = 2): string => {
  const output: string[] = [];
  for (const entry of traverseUnassignedFileInfo(fileInfo)) {
    const prefix = ' '.repeat(indent * entry.level);
    output.push(`${prefix}${entry.fileInfo.path}`);
  }

  return output.join(EOL);
};
