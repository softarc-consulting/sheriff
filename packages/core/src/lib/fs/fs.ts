import * as path from 'path';
import type { FsPath } from '../file-info/fs-path';

export abstract class Fs {
  abstract writeFile: (filename: string, contents: string) => void;
  abstract appendFile(filename: string, contents: string): void;
  abstract readFile: (path: FsPath) => string;
  abstract removeDir: (path: FsPath) => void;
  abstract createDir: (path: string) => void;
  abstract exists(path: string): path is FsPath;

  abstract tmpdir: () => string;

  join = (...paths: string[]) => path.join(...paths);

  abstract cwd: () => string;

  abstract findFiles: (path: FsPath, filename: string) => FsPath[];

  abstract print: () => void;

  /**
   * Used for finding the nearest `tsconfig.json`. It traverses through the
   * parent folder and includes the directory of the referenceFile.
   * @param referenceFile
   * @param filename
   */
  abstract findNearestParentFile: (
    referenceFile: FsPath,
    filename: string
  ) => FsPath;

  relativeTo = (from: string, to: string) => path.relative(from, to);

  getParent = (fileOrDirectory: FsPath): FsPath =>
    path.dirname(fileOrDirectory) as FsPath;

  pathSeparator = path.sep;

  /**
   * Reset the VirtualFs, has no effect on the real `DefaultFs`.
   */
  abstract reset(): void;

  abstract split(path: string): string[];

  abstract isAbsolute(path: string): boolean;
}
