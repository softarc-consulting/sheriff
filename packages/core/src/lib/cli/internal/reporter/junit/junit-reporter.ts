import { ProjectViolation } from '../../../project-violation';
import { cli } from '../../../cli';
import { Reporter } from '../reporter';
import { junitBuilder } from './internal/junit-report-builder';
import { saveReport } from '../utils/save-report';
import { ReporterOptions } from '../utils/reporter-options';

export class JunitReporter implements Reporter {
  #options: ReporterOptions;

  constructor(options: ReporterOptions) {
    this.#options = options;
  }

  createReport(validationResults: ProjectViolation): void {
    cli.log(`Creating JUnit report`);

    const xmlContent = this.#generateXml(validationResults);
    saveReport(this.#options, 'violations', '.xml', xmlContent);
  }

  #generateXml(validationResults: ProjectViolation): string {
    const builder = junitBuilder();
    const suite = builder.testsuite({
      name: this.#options.projectName,
      totalDependencyRulesViolations:
        validationResults.totalDependencyRuleViolations,
      totalEncapsulationViolations:
        validationResults.totalEncapsulationViolations,
      totalViolatedFiles: validationResults.totalViolatedFiles,
      hasError: validationResults.hasError,
    });

    // Process each violation
    for (const violation of validationResults.violations) {
      // Add encapsulation violations
      for (const encapsulation of violation.encapsulations) {
        suite.addTestCase({
          modulePath: violation.filePath,
          name: 'encapsulation',
          failureMessage: `${encapsulation} cannot be imported. It is encapsulated.`,
        });
      }

      // Add dependency rule violations
      for (const depViolation of violation.dependencyRuleViolations) {
        const fromModulePath = depViolation.fromModulePath;
        const toModulePath = depViolation.toModulePath;

        suite.addTestCase({
          modulePath: violation.filePath,
          name: 'dependency-rule',
          failureMessage: `module ${fromModulePath} cannot access ${toModulePath}. Tag ${depViolation.fromTag} has no clearance for tags ${depViolation.toTags.join(',')}`,
          fromTag: depViolation.fromTag,
          toTags: depViolation.toTags.join(','),
          fromModulePath: fromModulePath,
          toModulePath: toModulePath,
        });
      }
    }

    return builder.getReport();
  }
}
