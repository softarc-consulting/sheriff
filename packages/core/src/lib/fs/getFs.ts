import defaultFs from './default-fs';
import virtualFs from './virtual-fs';
import { Fs } from './fs';

let fsImplementation: 'default' | 'virtual' = 'default';

export const useDefaultFs = () => (fsImplementation = 'default');
export const useVirtualFs = () => (fsImplementation = 'virtual');

const getFs = (): Fs =>
  fsImplementation === 'default' ? defaultFs : virtualFs;

export const isFsVirtualised = () => fsImplementation === 'virtual';

export default getFs;
