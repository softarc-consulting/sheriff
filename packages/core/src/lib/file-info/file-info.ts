import { FsPath, toFsPath } from './fs-path';
import throwIfNull from '../util/throw-if-null';
import getFs from '../fs/getFs';

/**
 * Class representing a TypeScript file with its dependencies.
 * If an import cannot be resolved, it doesn't throw an error
 * but is added to unresolvableImports.
 *
 * It is up to the consumer, e.g. ESLinter, to decide if that
 * should cause an error or not.
 */
export default class FileInfo {
  #rawImportMap = new Map<string, string>();
  #unresolvableImports: string[] = [];
  constructor(public path: FsPath, public imports: FileInfo[] = []) {}

  addUnresolvableImport(importCommand: string) {
    this.#unresolvableImports.push(importCommand);
  }

  isUnresolvableImport(importCommand: string) {
    return this.#unresolvableImports.includes(importCommand);
  }

  hasUnresolvableImports() {
    return this.#unresolvableImports.length > 0;
  }

  addImport(importedFileInfo: FileInfo, rawImport: string) {
    this.imports.push(importedFileInfo);
    this.#rawImportMap.set(importedFileInfo.path, rawImport);
  }

  getRawImportForImportedFileInfo(path: FsPath): string {
    return throwIfNull(
      this.#rawImportMap.get(path),
      `raw import for ${path} is not available in ${this.path}`
    );
  }
}

type NestedArray = (string | NestedArray)[];

const createPath = (path: string, parentPath: string) => {
  let currentPath = path;
  if (path.startsWith('./')) {
    const parentPaths = parentPath.split('/');
    parentPaths.pop();
    const normalisedPaths = path.split('/').slice(1);

    currentPath = [...parentPaths, ...normalisedPaths].join('/');
  }

  return currentPath;
};

/**
 * utility function generate a FileInfo Tree for testing purposes
 * @param path
 * @param imports
 */
export const buildFileInfo = (
  path: string,
  imports: NestedArray = []
): FileInfo => {
  const fs = getFs();
  const children: FileInfo[] = imports.map((entry) => {
    if (
      Array.isArray(entry) &&
      entry.length === 2 &&
      typeof entry[0] === 'string'
    ) {
      const [childPath, childImports] = entry;
      const currentPath = createPath(childPath, path);
      return buildFileInfo(currentPath, childImports as NestedArray[]);
    } else if (typeof entry === 'string') {
      return buildFileInfo(createPath(entry, path), []);
    } else {
      throw new Error(`import not in right structure`);
    }
  });

  fs.writeFile(path, '');
  const fileInfo = new FileInfo(toFsPath(path));
  for (const child of children) {
    fileInfo.addImport(child, fs.relativeTo(fileInfo.path, child.path));
  }
  return fileInfo;
};
