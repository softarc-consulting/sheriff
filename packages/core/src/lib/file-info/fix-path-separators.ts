import { FsPath, toFsPath } from './fs-path';
import getFs from '../fs/getFs';

/**
 * Ensures that `FsPath` uses the separator from the OS and not always '/'
 *
 * @param path
 */
export function fixPathSeparators(path: string): FsPath {
  const fs = getFs();

  if (fs.pathSeparator !== '/') {
    return toFsPath(path.replace(/\//g, fs.pathSeparator));
  }

  return toFsPath(path);
}
