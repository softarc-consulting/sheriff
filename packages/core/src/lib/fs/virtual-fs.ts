import { Fs } from './fs';
import throwIfNull from '../util/throw-if-null';
import { FsPath, toFsPath } from '../file-info/fs-path';

type FsNodeType = 'file' | 'directory';

type FsNode = {
  parent: FsNode | undefined;
  children: Map<string, FsNode>;
  contents: string;
  type: FsNodeType;
  name: string;
};

type GetPathReturnSuccess = {
  exists: true;
  node: FsNode;
  parent: FsNode | undefined;
  nodeName: string;
};

type GetPathReturnFailure = {
  exists: false;
  node: undefined;
  parent: FsNode;
  restPaths: string[];
};

type GetPathReturn = GetPathReturnSuccess | GetPathReturnFailure;

export class VirtualFs extends Fs {
  // non-null assertion is used because Ts can't infer the init call
  root!: FsNode;
  project!: FsNode;

  constructor() {
    super();
    this.init();
  }

  init() {
    this.root = {
      parent: undefined,
      children: new Map<string, FsNode>(),
      contents: '',
      type: 'directory',
      name: 'root',
    };

    this.project = {
      parent: this.root,
      children: new Map<string, FsNode>(),
      contents: '',
      type: 'directory',
      name: 'project',
    };
    this.root.children.set('project', this.project);
  }

  findFiles = (path: FsPath, filename: string): FsPath[] => {
    const result = this.#getNode(path);
    if (!result.exists) {
      throw new Error(`directory ${path} does not exist`);
    } else if (result.node.type !== 'directory') {
      throw new Error(`${path} is not a directory`);
    }

    const referenceNode = result.node;
    return this.#traverseFindFiles(referenceNode, filename, referenceNode);
  };

  createDir = (path: string): void => {
    const node = this.#makeOrGet(path, 'directory');
    if (node.type === 'file') {
      throw new Error(`cannot create directory ${path} because it is a file`);
    }
  };

  override exists(path: string): path is FsPath {
    return this.#getNode(path).exists;
  }

  writeFile = (path: string, contents: string): void => {
    const node = this.#makeOrGet(path, 'file');
    if (node.type === 'directory') {
      throw new Error(`cannot write to file ${path} because it is a directory`);
    }
    node.contents = contents;
  };

  readFile = (path: string): string => {
    const result = this.#getNode(path);
    if (!result.exists) {
      throw new Error(`File ${path} does not exist`);
    } else if (result.node.type === 'directory') {
      throw new Error(
        `cannot read from file ${path} because it is a directory`
      );
    } else {
      return result.node.contents;
    }
  };

  removeDir = (path: string): void => {
    const result = this.#getNode(path);
    if (!result.exists) {
      throw new Error(
        `cannot delete directory ${path} because it does not exist`
      );
    }
    if (result.parent === undefined) {
      throw new Error(`cannot delete root directory`);
    }
    result.parent.children.delete(result.nodeName);
  };

  tmpdir = () => '/tmp';

  findNearestParentFile = (referenceFile: FsPath, filename: string): FsPath => {
    const { node } = this.#getNodeOrThrow(referenceFile);
    let current = throwIfNull(
      node.parent,
      `${referenceFile} must have a parent`
    );
    while (current.parent !== undefined) {
      const searchedNode = Array.from(current.children.values()).find(
        (childNode) => childNode.name === filename && childNode.type === 'file'
      );
      if (searchedNode) {
        return this.#absolutePath(searchedNode);
      }
      current = current.parent;
    }

    throw new Error(`cannot find ${filename} near ${referenceFile}`);
  };

  isAbsolute = (path: string) => path.startsWith('/');

  #makeOrGet = (path: string, type: FsNodeType): FsNode => {
    const result = this.#getNode(path);
    if (result.exists) {
      return result.node;
    }
    if (result.parent === undefined) {
      throw new Error(`could not determine root in ${path}`);
    }

    let node = result.parent;

    for (let i = 0; i < result.restPaths.length; i++) {
      const path = result.restPaths[i];
      const childNode = this.#createNode(
        node,
        i === result.restPaths.length - 1 ? type : 'directory',
        path
      );
      node.children.set(path, childNode);
      node = childNode;
    }

    return node;
  };

  #getNode = (path: string): GetPathReturn => {
    let node = path.startsWith('/') ? this.root : this.project;
    const paths = path.split('/').filter(Boolean);
    for (let i = 0; i < paths.length; i++) {
      const path = paths[i];
      if (path === '..') {
        if (node.parent === undefined) {
          throw new Error('/ has no parent');
        }
        node = node.parent;
      } else if (path !== '.') {
        const childNode = node.children.get(path);
        if (childNode) {
          node = childNode;
        } else {
          return {
            exists: false,
            node: undefined,
            parent: node,
            restPaths: paths.slice(i),
          };
        }
      }
    }

    return {
      exists: true,
      node,
      parent: node.parent,
      nodeName: paths.at(-1) || '',
    };
  };

  #getNodeOrThrow = (path: string): GetPathReturnSuccess => {
    const response = this.#getNode(path);
    if (!response.exists) {
      throw new Error(`file ${path} does not exist`);
    }

    return response;
  };

  #createNode(parent: FsNode, type: FsNodeType, name: string): FsNode {
    return {
      parent,
      children: new Map<string, FsNode>(),
      contents: '',
      type,
      name,
    };
  }

  #traverseFindFiles = (
    node: FsNode,
    filename: string,
    referenceNode: FsNode
  ): FsPath[] => {
    const found: FsPath[] = [];
    for (const childNode of node.children.values()) {
      if (
        childNode.type === 'file' &&
        childNode.name === filename.toLowerCase()
      ) {
        found.push(this.#absolutePath(childNode));
      }

      if (childNode.type === 'directory') {
        found.push(
          ...this.#traverseFindFiles(childNode, filename, referenceNode)
        );
      }
    }

    return found;
  };

  #absolutePath = (node: FsNode): FsPath => {
    const paths = [];
    let current = node;
    while (current.parent !== undefined && current !== this.root) {
      paths.push(current.name);
      current = current.parent;
    }

    return toFsPath('/' + paths.reverse().join('/'));
  };

  print = (node?: FsNode, indent = 0): void => {
    if (node === undefined) {
      console.log('[root]');
      return this.print(this.root, indent + 2);
    }

    for (const child of node.children.keys()) {
      const childNode = node.children.get(child);
      console.log(
        `${' '.repeat(indent)}${
          childNode?.type === 'directory' ? '[' + child + ']' : child
        }`
      );
      this.print(node.children.get(child), indent + 2);
    }
  };

  cwd = () => '/project';

  reset() {
    this.init();
  }

  override split(path: string): string[] {
    return path.split('/');
  }
}

const virtualFs = new VirtualFs();
export default virtualFs;
