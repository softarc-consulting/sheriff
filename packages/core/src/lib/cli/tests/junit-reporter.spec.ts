import { beforeEach, describe, it, expect, beforeAll } from 'vitest';
import { VirtualFs } from '../../fs/virtual-fs';
import getFs, { useVirtualFs } from '../../fs/getFs';
import { toFsPath } from '../../file-info/fs-path';
import { ProjectViolation } from '../project-violation';
import { JunitReporter } from '../internal/reporter/junit/junit-reporter';

describe('JUnit reporter', () => {
  let fs: VirtualFs;
  beforeAll(() => {
    useVirtualFs();
    fs = getFs() as VirtualFs;
  });

  beforeEach(() => {
    fs.reset();
    fs.createDir('/project/customers/feature');
    fs.createDir('/project/shared/master-data');
    fs.createDir('/project/customers/ui');
    fs.createDir('/project/app/shared/form');
  });
  it('should create a xml-file in /.sheriff/project/violations.xml', () => {
    const reporter = new JunitReporter({
      outputDir: '.sheriff',
      projectName: 'project',
    });
    const violations: ProjectViolation = {
      totalEncapsulationViolations: 0,
      totalViolatedFiles: 0,
      totalDependencyRuleViolations: 2,
      hasError: true,
      violations: [
        {
          filePath: '/project/customers/feature/feature.ts',
          encapsulations: [],
          dependencyRules: [],
          dependencyRuleViolations: [
            {
              rawImport: '@eternal/shared/master-data',
              fromModulePath: toFsPath('/project/customers/feature'),
              toModulePath: toFsPath('/project/shared/master-data'),
              fromTag: 'domain:customers',
              toTags: ['shared:master-data'],
            },
          ],
        },
        {
          filePath: '/project/customers/ui/ui.ts',
          encapsulations: [],
          dependencyRules: [],
          dependencyRuleViolations: [
            {
              rawImport: '@eternal/shared/form',
              fromModulePath: toFsPath('/project/customers/ui'),
              toModulePath: toFsPath('/project/app/shared/form'),
              fromTag: 'domain:customers',
              toTags: ['shared:form'],
            },
          ],
        },
      ],
    };
    reporter.createReport(violations);

    expect(fs.readFile('.sheriff/project/violations.xml')).toMatchSnapshot();
  });
});
