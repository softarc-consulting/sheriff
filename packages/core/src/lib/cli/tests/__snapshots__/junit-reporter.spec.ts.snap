// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`JUnit reporter > should create a xml-file in /.sheriff/project/violations.xml 1`] = `
"<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
  <testsuite name="project" totalDependencyRulesViolations="2" totalEncapsulationViolations="0" totalViolatedFiles="0" hasError="true">
    <testcase modulePath="/project/customers/feature/feature.ts" name="dependency-rule" fromTag="domain:customers" toTags="shared:master-data" fromModulePath="/project/customers/feature" toModulePath="/project/shared/master-data">
      <failure message="module /project/customers/feature cannot access /project/shared/master-data. Tag domain:customers has no clearance for tags shared:master-data"/>
    </testcase>
    <testcase modulePath="/project/customers/ui/ui.ts" name="dependency-rule" fromTag="domain:customers" toTags="shared:form" fromModulePath="/project/customers/ui" toModulePath="/project/app/shared/form">
      <failure message="module /project/customers/ui cannot access /project/app/shared/form. Tag domain:customers has no clearance for tags shared:form"/>
    </testcase>
  </testsuite>
</testsuites>"
`;
